runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/state/StateRequestHandlers.java

```diff
@@ -0,0 +1,270 @@
+            Coder<W> windowCoder) {
+          throw new UnsupportedOperationException(String.format(
+              "The %s does not support handling sides inputs for PTransform %s with user state "
+                  + "id %s.",
+              BagUserStateHandler.class.getSimpleName(),
+              pTransformId,
+              userStateId));
+        }
+      };
+    }
+  }
+
+  /**
+   * Returns an adapter which converts a {@link MultimapSideInputHandlerFactory} to
+   * {@link StateRequestHandler}.
+   *
+   * <p>The {@link MultimapSideInputHandlerFactory} is required to handle all multimap side inputs
+   * contained within the {@link ExecutableProcessBundleDescriptor}. See
+   * {@link ExecutableProcessBundleDescriptor#getMultimapSideInputSpecs} for the set of multimap
+   * side inputs that are contained.
+   */
+  public static StateRequestHandler forMultimapSideInputHandlerFactory(
+      ExecutableProcessBundleDescriptor processBundleDescriptor,
+      MultimapSideInputHandlerFactory multimapSideInputHandlerFactory) {
+    return new StateRequestHandlerToMultimapSideInputHandlerFactoryAdapter(
+        processBundleDescriptor, multimapSideInputHandlerFactory);
+  }
+
+  /**
+   * An adapter which converts {@link MultimapSideInputHandlerFactory} to
+   * {@link StateRequestHandler}.
+   */
+  static class StateRequestHandlerToMultimapSideInputHandlerFactoryAdapter
+      implements StateRequestHandler {
+
+    private final ExecutableProcessBundleDescriptor processBundleDescriptor;
+    private final MultimapSideInputHandlerFactory multimapSideInputHandlerFactory;
+    private final ConcurrentHashMap<MultimapSideInputSpec, MultimapSideInputHandler> cache;
+
+    StateRequestHandlerToMultimapSideInputHandlerFactoryAdapter(
+        ExecutableProcessBundleDescriptor processBundleDescriptor,
+        MultimapSideInputHandlerFactory multimapSideInputHandlerFactory) {
+      this.processBundleDescriptor = processBundleDescriptor;
+      this.multimapSideInputHandlerFactory = multimapSideInputHandlerFactory;
+      this.cache = new ConcurrentHashMap<>();
+    }
+
+    @Override
+    public CompletionStage<StateResponse.Builder> handle(
+        StateRequest request) throws Exception {
+      try {
+        checkState(TypeCase.MULTIMAP_SIDE_INPUT.equals(request.getStateKey().getTypeCase()),
+            "Unsupported %s type %s, expected %s",
+            StateRequest.class.getSimpleName(),
+            request.getStateKey().getTypeCase(),
+            TypeCase.MULTIMAP_SIDE_INPUT);
+
+        StateKey.MultimapSideInput stateKey = request.getStateKey().getMultimapSideInput();
+        MultimapSideInputSpec<?, ?, ?> sideInputReferenceSpec =
+            processBundleDescriptor.getMultimapSideInputSpecs()
+                .get(stateKey.getPtransformId())
+                .get(stateKey.getSideInputId());
+        MultimapSideInputHandler<?, ?, ?> handler = cache.computeIfAbsent(
```

Comment:
Note that in general, side input handlers may have limited lifetimes and will be invalid if referenced outside of that lifetime. Should we provide a way to explicitly remove entries from the cache and/or mark them as inoperable?

In either case, please document the fact that the `MultimapSideInputHandler`s returned by factories must be thread-safe.

Reply:
Ah, that makes sense. I wasn't sure who was meant to actually construct the handlers themselves, but it makes sense if we keep that as it is now and have the runner directly call into this when creating handlers.